CBinding/CLang/CLangManager.cs

```diff
@@ -0,0 +1,482 @@
+﻿using System;
+using CBinding;
+using ClangSharp;
+using MonoDevelop.Ide;
+using MonoDevelop.Projects;
+using System.Collections.Generic;
+using System.Threading;
+using MonoDevelop.Ide.Editor;
+using System.Runtime.InteropServices;
+using CBinding.Refactoring;
+using CBinding.Parser;
+using System.IO;
+
+namespace CBinding
+{
+	/// <summary>
+	/// Class to manage clang Translation units, thread safe
+	/// </summary>
+	public class CLangManager : IDisposable
+	{
+		public readonly object syncroot = new object ();
+		private CProject project;
+		private CXIndex index;
+		private Dictionary<string, CXTranslationUnit> translationUnits;
+
+		public Dictionary<string, CXTranslationUnit> TranslationUnits {
+			get {				
+				lock (syncroot) {
```

Comment:
I could find any information about libclang's internal threading solutions, the best i could find were some stackoverflow topic and a short mailing list discussion, where only assumptions were made. This lock is needed because clangmanager handles getting cursor, cursor references and cursor between files sometimes are a mess and can only identified by its USR, and e.g. finding a cursor by its USR while reparsing is in progress in a file could be a fault.